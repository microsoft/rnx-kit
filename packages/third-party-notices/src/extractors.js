/**
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * browserify-licenses v1.4.7-1 - License BSD-3-Clause
 * downloaded from <https://github.com/wanadev/browserify-licenses>
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Copyright (c) 2016, Wanadev
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright notice,
 *       this list of conditions and the following disclaimer in the documentation
 *       and/or other materials provided with the distribution.
 *     * Neither the name of browserify-licenses nor the names of its contributors
 *       may be used to endorse or promote products derived from this software
 *       without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
"use strict";

const path = require("path");
const fs = require("fs");
const { readPackage } = require("@rnx-kit/tools-node/package");
const spdxParse = require("spdx-expression-parse");

function _getPackageJsonInformations(modules) {
  for (let i = 0; i < modules.length; i++) {
    let module = modules[i];
    let pkg = readPackage(module.path);

    module.version = pkg.version;
    if (
      module.version.match(
        /^https?:\/\/.+-([0-9]+\.[0-9]+\.[0-9]+(-[a-z0-9._-]+)?)\.tgz$/i
      )
    ) {
      module.version = module.version.replace(
        /^https?:\/\/.+-([0-9]+\.[0-9]+\.[0-9]+(-[a-z0-9._-]+)?)\.tgz$/i,
        "$1"
      );
    }

    module.licenseURLs = [];
    if (pkg.license && typeof pkg.license == "string") {
      module.license = pkg.license;
      _getSpdxLicenseInformation(module.license, module.name).forEach(
        (license) => {
          module.licenseURLs.push(
            "https://spdx.org/licenses/" + license + ".html"
          );
        }
      );
    } else if (
      (pkg.license && Array.isArray(pkg.license)) ||
      (pkg.licenses && Array.isArray(pkg.licenses))
    ) {
      let licenses = pkg.license ? pkg.license : pkg.licenses;
      let licenseList = [];
      for (var j = 0; j < licenses.length; j++) {
        licenseList.push(licenses[j].type || licenses[j].name);
        module.license = licenseList.join(" / ");
        if (licenses[j].url) {
          module.licenseURLs.push(licenses[j].url);
        }
      }
    } else if (pkg.license && typeof pkg.license == "object") {
      module.license = pkg.license.type || pkg.license.name;
      if (pkg.license.url) {
        module.licenseURLs.push(pkg.license.url);
      }
    }
  }
  return modules;
}

function _findLicenseFiles(modules) {
  for (let i = 0; i < modules.length; i++) {
    let module = modules[i];
    var files = fs.readdirSync(module.path);

    for (var j = 0; j < files.length; j++) {
      if (files[j].match(/(LICENSE|LICENCE|COPYING)/i)) {
        module.licenseFile = files[j];
        if (module.noticeFile) break;
      } else if (files[j].match(/NOTICE/i)) {
        module.noticeFile = files[j];
        if (module.licenseFile) break;
      }
    }
  }
  return modules;
}

function _getSpdxLicenseInformation(license, moduleName) {
  var licenses = [];
  try {
    var tree;
    if (typeof license === "string") {
      tree = spdxParse(license);
    } else {
      tree = license;
    }
    if (tree.license) {
      licenses.push(tree.license);
    }
    if (tree.left) {
      licenses = licenses.concat(_getSpdxLicenseInformation(tree.left));
    }
    if (tree.right) {
      licenses = licenses.concat(_getSpdxLicenseInformation(tree.right));
    }
  } catch (e) {
    if (license.toUpperCase() !== "UNLICENSED") {
      console.warn(
        `WARNING: Unable to parse license "${license}" in ${moduleName}`
      );
    }
  }
  return licenses;
}

function _readLicenseText(modules) {
  for (let i = 0; i < modules.length; i++) {
    let module = modules[i];
    if (module.licenseFile) {
      module.licenseText = fs
        .readFileSync(path.join(module.path, module.licenseFile))
        .toString()
        .replace(/\r?\n/g, "\n");
    }
    if (module.noticeFile) {
      module.noticeText = fs
        .readFileSync(path.join(module.path, module.noticeFile))
        .toString()
        .replace(/\r?\n/g, "\n");
    }
  }
  return modules;
}

function extractorNodeModule(modules) {
  _getPackageJsonInformations(modules);
  _findLicenseFiles(modules);
  _readLicenseText(modules);
  return modules;
}

module.exports = {
  nodeModule: extractorNodeModule,
};
