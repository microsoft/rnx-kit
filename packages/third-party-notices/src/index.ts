export type {
  WriteThirdPartyNoticesOptions as IWriteThirdPartyNoticesOptions,
  WriteThirdPartyNoticesOptions,
} from "./types";
export { writeThirdPartyNotices } from "./write-third-party-notices";
