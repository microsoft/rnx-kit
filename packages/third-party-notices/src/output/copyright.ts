import { findPackage, readPackage } from "@rnx-kit/tools-node/package";

export function getPackageAuthor(modulePath: string): string | undefined {
  const pkgFile = findPackage(modulePath);
  if (pkgFile) {
    const manifest = readPackage(pkgFile);
    if (manifest) {
      return typeof manifest.author === "string"
        ? manifest.author
        : manifest.author?.name;
    }
  }
  return undefined;
}

export function parseCopyright(
  modulePath: string,
  licenseText: string | undefined,
  license: string | undefined,
  licenseURLs: string[]
): string {
  const m = licenseText?.match(/^Copyright .*$/m);
  if (!m) {
    const packageAuthor = getPackageAuthor(modulePath);
    if (packageAuthor) {
      return packageAuthor;
    }

    if (licenseURLs?.length > 0) {
      return `${license} (${licenseURLs.join(" ")})`;
    }

    return "No copyright notice";
  }

  return m[0].trim();
}
