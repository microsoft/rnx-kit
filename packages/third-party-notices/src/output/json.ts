import { findPackage, readPackage } from "@rnx-kit/tools-node/package";
import type { License } from "../types";

function getPackageAuthor(modulePath: string): string | undefined {
  const pkgFile = findPackage(modulePath);
  if (pkgFile) {
    const manifest = readPackage(pkgFile);
    if (manifest) {
      return typeof manifest.author === "string"
        ? manifest.author
        : manifest.author?.name;
    }
  }
  return undefined;
}

function parseCopyright(
  modulePath: string,
  licenseText: string | undefined
): string {
  const m = licenseText?.match(/^Copyright .*$/m);
  if (!m) {
    return getPackageAuthor(modulePath) || "No copyright notice";
  }

  return m[0].trim();
}

export function createLicenseJSON(licenses: License[]): string {
  return JSON.stringify({
    packages: licenses.map(
      ({ name, path: modulePath, version, license, licenseText }) => {
        if (!license) {
          throw new Error(`No license for ${name}`);
        }
        return {
          name,
          version,
          license,
          copyright: parseCopyright(modulePath, licenseText),
        };
      }
    ),
  });
}
