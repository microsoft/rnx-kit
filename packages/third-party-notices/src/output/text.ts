import os from "os";
import type { License } from "../types";

export function createLicenseFileContents(
  licenses: License[],
  preambleText?: string[],
  additionalText?: string[]
): string {
  let outputText = "";

  const writeLine = (s: string): void => {
    outputText += `${s || ""}${os.EOL}`;
  };

  const writeMultipleLines = (s: string): void => {
    const lines = s.split(/\r\n|\r|\n/g);
    lines.forEach((line: string) => {
      writeLine(line);
    });
  };

  if (preambleText) {
    writeMultipleLines(preambleText.join(os.EOL));
  }

  // Emit combined license text
  licenses.forEach(({ name, version, license, licenseText, licenseURLs }) => {
    if (license?.toUpperCase() === "UNLICENSED") {
      // Ignore unlicensed/private packages
      return;
    }

    if (!licenseText) {
      if (!license && (!licenseURLs || licenseURLs.length === 0)) {
        throw new Error(
          `No license information found for package '${name}'. Consider filing an issue for the project to properly advertise its licence. Pass this module to the tool via '--ignoreModules ${name}' to suppress this message.`
        );
      }
      licenseText = `${license} (${licenseURLs.join(" ")})`;
    }
    writeLine("================================================");
    writeLine(`${name} ${version}`);
    writeLine("--");
    writeMultipleLines(licenseText.trim());
    writeLine("================================================");
    writeLine("");
  });

  if (additionalText) {
    writeMultipleLines(additionalText.join(os.EOL));
  }

  return outputText;
}
