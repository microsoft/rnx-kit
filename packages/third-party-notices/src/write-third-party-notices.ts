import { isPackageModuleRef, parseModuleRef } from "@rnx-kit/tools-node/module";
import { findPackage, readPackage } from "@rnx-kit/tools-node/package";
import * as fs from "fs";
import type { BasicSourceMap } from "metro-source-map";
import * as path from "path";
import { createLicenseJSON } from "./output/json";
import { createLicenseFileContents } from "./output/text";
import type {
  License,
  ModuleNamePathPair,
  SourceMap,
  SourceSection,
  WriteThirdPartyNoticesOptions,
} from "./types";

const modulesRoot = "node_modules/";

/**
 * Function to write third party notices based on the specified source map.
 *
 * This function will read the sourcemap file and tries to find all files
 * that are referenced in the sourcemap by assuming that all dependencies are
 * represented as `node_modules\moduleName` or `node_modules\@scope\moduleName`
 * It will then look in the package.json file to see if it finds a license declaration
 * or it will look for the file called `LICENCE` or `LICENSE` in the root. And aggregate all these
 * files in the output file.
 *
 * This function works for npm, yarn and pnpm package layouts formats.
 *
 * At the moment this function only supports webpack based bundles, there is nothing
 * preventing adding metro support, the current customers of this module are based on
 * webpack at the moment.
 *
 * @param options See IWriteThirdPartyNoticesOptions for more details
 */
export async function writeThirdPartyNotices(
  options: WriteThirdPartyNoticesOptions
): Promise<void> {
  const fileOptions = { encoding: "utf-8" } as const;
  const { additionalText, json, outputFile, preambleText, sourceMapFile } =
    options;

  // Parse source map file
  const sourceMapJson = fs.readFileSync(sourceMapFile, fileOptions);
  const sourceMap: SourceMap = JSON.parse(sourceMapJson);

  const currentPackageId = await getCurrentPackageId(options.rootPath);

  const moduleNameToPathMap = extractModuleNameToPathMap(
    options,
    currentPackageId,
    sourceMap
  );
  const licenses = await extractLicenses(moduleNameToPathMap);
  const outputText = json
    ? createLicenseJSON(licenses)
    : createLicenseFileContents(licenses, preambleText, additionalText);

  if (outputFile) {
    fs.writeFileSync(outputFile, outputText, fileOptions);
  } else {
    console.log(outputText);
  }
}

export async function writeThirdPartyNoticesFromMap(
  options: WriteThirdPartyNoticesOptions,
  moduleNameToPathMap: Map<string, string>
): Promise<void> {
  const { additionalText, json, preambleText, sourceMapFile } = options;
  let { outputFile } = options;

  const licenses = await extractLicenses(moduleNameToPathMap);
  const outputText = json
    ? createLicenseJSON(licenses)
    : createLicenseFileContents(licenses, preambleText, additionalText);

  if (!outputFile) {
    const prefix = sourceMapFile?.replace(/\.map$/, "") ?? "main";
    outputFile = prefix + ".LICENSE.txt";
  }

  fs.writeFileSync(outputFile, outputText);
}

export async function getCurrentPackageId(
  rootPath: string
): Promise<string | undefined> {
  const pkgFile = findPackage(rootPath);
  if (pkgFile) {
    const manifest = readPackage(pkgFile);
    return manifest?.name;
  }

  return undefined;
}

// helper functions
export function normalizePath(p: string, currentPackageId?: string): string {
  let result = p.replace(/webpack:\/\/\//g, "");
  if (currentPackageId) {
    result = result.replace(
      new RegExp(`webpack://${currentPackageId}/`, "g"),
      ""
    );
  }
  return result.replace(/[\\]+/g, "/");
}

export function splitSourcePath(rootPath: string, p: string): string[] {
  //  find the root of the package, and extract the package name.
  //
  //  npm packages are laid out simply. pnpm packages have additional layers
  //  of hierarchy in the package store, such as package url and version.
  //
  //  ** example **
  //    npm:  <repo-root>/common/temp/node_modules/fbjs
  //    pnpm: <repo-root>/common/temp/node_modules/.yourProject.pkgs.visualstudio.com/fbjs/0.8.17/node_modules/fbjs
  //
  //  bundles use WebPack's dll plugin to perform cross-bundle module retrieval.
  //  cross bundle modules' paths are appended with 'delegated ', which we need
  //  to remove from path.
  //
  //  ** example **
  //    delegated ../../../common/temp/node_modules/.yourProject.pkgs.visualstudio.com/react-native/0.3003.5/react@16.0.0/node_modules/react-native/Libraries/EventEmitter/NativeEventEmitter.js from dll-reference runtime20
  const DELEGATED_PREFIX = "delegated ";
  if (p.startsWith(DELEGATED_PREFIX)) {
    p = p.substring(DELEGATED_PREFIX.length);
  }

  const absolutePath = path.resolve(rootPath, p).replace(/[\\]+/g, "/");
  const idx = absolutePath.lastIndexOf(modulesRoot);
  if (idx == null || idx < 0) {
    throw new Error(`Unexpected module: ${p}`);
  }

  const ref = parseModuleRef(absolutePath.substring(idx + modulesRoot.length));
  if (!isPackageModuleRef(ref)) {
    throw new Error(`Could not parse module: ${p}`);
  }

  return [
    ref.scope ? `${ref.scope}/${ref.name}` : ref.name,
    ref.path
      ? absolutePath.substring(0, absolutePath.length - ref.path.length - 1)
      : absolutePath,
  ];
}

export function parseModule(
  options: WriteThirdPartyNoticesOptions,
  moduleNameToPath: Map<string, string>,
  p: string
): void {
  const [moduleName, modulePath] = splitSourcePath(options.rootPath, p);
  if (
    (options.ignoreScopes &&
      options.ignoreScopes.some((scope: string) =>
        moduleName.startsWith(scope + "/")
      )) ||
    (options.ignoreModules &&
      options.ignoreModules.indexOf(moduleName) !== -1) ||
    !fs.existsSync(modulePath)
  ) {
    return;
  }

  moduleNameToPath.set(moduleName, modulePath);
}

export function parseSourceMap(
  options: WriteThirdPartyNoticesOptions,
  currentPackageId: string | undefined,
  moduleNameToPath: Map<string, string>,
  sourceMap: SourceMap
): void {
  sourceMap.sources.forEach((source: string) => {
    source = normalizePath(source, currentPackageId);
    if (source.includes(modulesRoot)) {
      parseModule(options, moduleNameToPath, source);
    }
  });
}

export function extractModuleNameToPathMap(
  options: WriteThirdPartyNoticesOptions,
  currentPackageId: string | undefined,
  sourceMap: SourceMap
): Map<string, string> {
  const moduleNameToPathMap = new Map<string, string>();

  if (sourceMap.sources) {
    parseSourceMap(options, currentPackageId, moduleNameToPathMap, sourceMap);
  }
  if (sourceMap.sections) {
    sourceMap.sections.forEach((section: SourceSection) => {
      parseSourceMap(
        options,
        currentPackageId,
        moduleNameToPathMap,
        section.map
      );
    });
  }

  return moduleNameToPathMap;
}

export async function extractLicenses(
  moduleNameToPathMap: Map<string, string>
): Promise<License[]> {
  const licenseExtractors = require("./extractors");

  const moduleNamePathPairs: ModuleNamePathPair[] = [];
  moduleNameToPathMap.forEach((modulePath: string, moduleName: string) => {
    // If both foo and @foo/bar exist, only include the license for foo
    if (moduleName[0] === "@") {
      const parentModuleName = moduleName.split("/")[0].substring(1);
      if (moduleNameToPathMap.has(parentModuleName)) {
        moduleNameToPathMap.delete(moduleName);
        return;
      }
    }

    moduleNamePathPairs.push({
      name: moduleName,
      path: modulePath,
    });
  });

  // Extract licenses of all modules we found
  return await licenseExtractors.nodeModule(
    moduleNamePathPairs.sort(({ name: lhs }, { name: rhs }) =>
      lhs < rhs ? -1 : lhs > rhs ? 1 : 0
    )
  );
}

export function gatherModulesFromSources(
  sources: BasicSourceMap["sources"],
  options: WriteThirdPartyNoticesOptions
): Map<string, string> {
  const moduleNameToPath = new Map<string, string>();

  sources.forEach((source) => {
    if (source.includes("node_modules")) {
      source = normalizePath(source);
      parseModule(options, moduleNameToPath, source);
    }
  });

  return moduleNameToPath;
}
