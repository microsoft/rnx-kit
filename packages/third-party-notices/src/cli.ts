import { error } from "@rnx-kit/console";
import * as path from "path";
import * as yargs from "yargs";
import type { WriteThirdPartyNoticesOptions } from "./types";
import { writeThirdPartyNotices } from "./write-third-party-notices";

function fail(message: string): never {
  error(message);
  process.exit(1);
}

function getArgs(): WriteThirdPartyNoticesOptions {
  const argv = yargs.options({
    rootPath: {
      type: "string",
      describe: "The root of the repo where to start resolving modules from.",
      require: true,
    },
    sourceMapFile: {
      type: "string",
      describe: "The sourceMap file to generate license contents for.",
      require: true,
    },
    json: {
      type: "boolean",
      describe: "Output license information as a JSON",
      default: false,
    },
    outputFile: {
      type: "string",
      describe: "The output file to write the license file to.",
    },
    ignoreScopes: {
      string: true,
      array: true,
      describe: "Npm scopes to ignore and not emit license information for",
    },
    ignoreModules: {
      string: true,
      array: true,
      describe: "Modules (js packages) to not emit license information for ",
    },
    preambleText: {
      string: true,
      array: true,
      describe:
        "A list of lines to prepend at the start of the generated license file.",
    },
    additionalText: {
      string: true,
      array: true,
      describe:
        "A list of lines to append at the end of the generated license file.",
    },
  }).argv;

  const writeTpnArgs: WriteThirdPartyNoticesOptions = argv;

  if (!writeTpnArgs.outputFile) {
    writeTpnArgs.outputFile =
      writeTpnArgs.sourceMapFile?.substring(
        0,
        writeTpnArgs.sourceMapFile.length -
          path.extname(writeTpnArgs.sourceMapFile).length
      ) + ".tpn";
  }

  return writeTpnArgs;
}

async function main(): Promise<void> {
  const writeTpnArgs = getArgs();
  await writeThirdPartyNotices(writeTpnArgs);
}

main().catch((err: string) => fail(err));
