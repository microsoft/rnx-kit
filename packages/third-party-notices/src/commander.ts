import type { Config } from "@react-native-community/cli-types";
import { cliOptions } from "./cli";
import type { WriteThirdPartyNoticesOptions } from "./types";
import { writeThirdPartyNotices } from "./write-third-party-notices";

type InputArgs = WriteThirdPartyNoticesOptions & {
  ignoreScopes?: string;
  ignoreModules?: string;
  preambleText?: string;
  additionalText?: string;
};

function rnxWriteThirdPartyNotices(
  _argv: string[],
  _config: Config,
  {
    additionalText,
    fullLicenseText,
    ignoreModules,
    ignoreScopes,
    json,
    outputFile,
    preambleText,
    rootPath,
    sourceMapFile,
  }: InputArgs
): void {
  // react-native-cli is not as rich as yargs, so we have to perform a mapping.
  writeThirdPartyNotices({
    rootPath,
    sourceMapFile,
    json,
    outputFile,
    ignoreScopes: ignoreScopes?.split(","),
    ignoreModules: ignoreModules?.split(","),
    preambleText: preambleText ? [preambleText] : undefined,
    additionalText: additionalText ? [additionalText] : undefined,
    fullLicenseText,
  });
}

export const writeThirdPartyNoticesCommand = {
  name: "rnx-write-third-party-notices",
  description: "Writes third party notices based on the given bundle",
  func: rnxWriteThirdPartyNotices,
  get options() {
    return Object.entries(cliOptions).map(([flag, options]) => {
      return {
        name:
          "argsString" in options
            ? `--${flag} ${options.argsString}`
            : `--${flag}`,
        description: options.description,
        default: "default" in options ? options.default : undefined,
      };
    });
  },
};
