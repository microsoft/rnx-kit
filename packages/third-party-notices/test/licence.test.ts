import path from "path";
import { createLicenseJSON } from "../src/output/json";
import { createLicenseFileContents } from "../src/output/text";
import type { License } from "../src/types";
import { extractLicenses } from "../src/write-third-party-notices";

async function getSampleLicenseData(): Promise<License[]> {
  const map = new Map([
    // License data in package.json
    [
      "@rnx-kit/console",
      path.dirname(require.resolve("@rnx-kit/console/package.json")),
    ],
    // License data package.json and LICENCE file
    ["yargs", path.dirname(require.resolve("yargs/package.json"))],
  ]);

  const licenses = await extractLicenses(map);

  // Hack versions to not depend on
  for (const license of licenses) {
    license.version = "1.2.3-fixedVersionForTesting";
  }

  // Private packages should be excluded from text output
  licenses.push({
    name: "private-package",
    version: "1.0.0",
    license: "Unlicensed",
    licenseURLs: [],
    path: ".",
  });

  return licenses;
}

describe("license", () => {
  test("extractLicenses", async () => {
    const licenses = await getSampleLicenseData();

    // normalize the paths for stable and cross platform snapshots
    for (const license of licenses) {
      license.path = license.path
        .replace(path.resolve(__dirname, "../../.."), "~")
        .replace(/[/\\]/g, "/");
    }

    expect(licenses).toMatchSnapshot();
  });

  test("createLicenseFileContents", async () => {
    const licenses = await getSampleLicenseData();

    const licenseText = createLicenseFileContents(licenses);

    expect(licenseText).toMatchSnapshot();
  });

  test("createLicenseFileContentsWithWrappers", async () => {
    const licenses = await getSampleLicenseData();

    const licenseText = createLicenseFileContents(
      licenses,
      ["preamble 1\n2\r\n3\r4", "Preamble 2"],
      ["additional 1\n2\r\n3\r4", "additional 2"]
    );

    expect(licenseText).toMatchSnapshot();
  });

  test("createLicenseJSON", async () => {
    const licenses = await getSampleLicenseData();

    const licenseText = createLicenseJSON(licenses);

    expect(JSON.parse(licenseText)).toMatchSnapshot();
  });
});
