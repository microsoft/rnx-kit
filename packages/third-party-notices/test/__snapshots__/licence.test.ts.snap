// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`license createLicenseFileContents 1`] = `
"================================================
@rnx-kit/console 1.2.3-fixedVersionForTesting
--
MIT (https://spdx.org/licenses/MIT.html)
================================================

================================================
yargs 1.2.3-fixedVersionForTesting
--
MIT License

Copyright 2010 James Halliday (mail@substack.net); Modified work Copyright 2014 Contributors (ben@npmjs.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the \\"Software\\"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED \\"AS IS\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
================================================

"
`;

exports[`license createLicenseFileContentsWithWrappers 1`] = `
"preamble 1
2
3
4
Preamble 2
================================================
@rnx-kit/console 1.2.3-fixedVersionForTesting
--
MIT (https://spdx.org/licenses/MIT.html)
================================================

================================================
yargs 1.2.3-fixedVersionForTesting
--
MIT License

Copyright 2010 James Halliday (mail@substack.net); Modified work Copyright 2014 Contributors (ben@npmjs.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the \\"Software\\"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED \\"AS IS\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
================================================

additional 1
2
3
4
additional 2
"
`;

exports[`license createLicenseJSON 1`] = `
Object {
  "packages": Array [
    Object {
      "copyright": "No copyright notice",
      "license": "MIT",
      "name": "@rnx-kit/console",
      "version": "1.2.3-fixedVersionForTesting",
    },
    Object {
      "copyright": "Copyright 2010 James Halliday (mail@substack.net); Modified work Copyright 2014 Contributors (ben@npmjs.com)",
      "license": "MIT",
      "name": "yargs",
      "version": "1.2.3-fixedVersionForTesting",
    },
    Object {
      "copyright": "No copyright notice",
      "license": "Unlicensed",
      "name": "private-package",
      "version": "1.0.0",
    },
  ],
}
`;

exports[`license extractLicenses 1`] = `
Array [
  Object {
    "license": "MIT",
    "licenseURLs": Array [
      "https://spdx.org/licenses/MIT.html",
    ],
    "name": "@rnx-kit/console",
    "path": "~/packages/console",
    "version": "1.2.3-fixedVersionForTesting",
  },
  Object {
    "license": "MIT",
    "licenseFile": "LICENSE",
    "licenseText": "MIT License

Copyright 2010 James Halliday (mail@substack.net); Modified work Copyright 2014 Contributors (ben@npmjs.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the \\"Software\\"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED \\"AS IS\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
",
    "licenseURLs": Array [
      "https://spdx.org/licenses/MIT.html",
    ],
    "name": "yargs",
    "path": "~/node_modules/yargs",
    "version": "1.2.3-fixedVersionForTesting",
  },
  Object {
    "license": "Unlicensed",
    "licenseURLs": Array [],
    "name": "private-package",
    "path": ".",
    "version": "1.0.0",
  },
]
`;
